analysis/pseudobulk-bulk-prediction/notebooks/perform-ora.Rmd

```diff
@@ -0,0 +1,270 @@
+
+```{r}
+baseline_or_df
+```
+
+## Permutation test 
+
+To calculate P-values here, we shuffle the residual/gene_id relationship and perform a permutation test.
+
+```{r}
+shuffled_or_df <- 1:params$reps |>
+  purrr::set_names(1:params$reps) |>
+  purrr::map(\(i) {
+    
+    shuffled_df <- data.frame(
+      ensembl_id = summary_residual_df$ensembl_id, 
+      summary_residual = sample(summary_residual_df$summary_residual, nrow(summary_residual_df))   
+    )
+    
+    geneset_list |>
+      purrr::map(
+        \(gs_df) analyze_outliers(shuffled_df, gs_df$ensembl_id)
+      ) |>
+      purrr::list_rbind(names_to = "geneset_cell_type") |>
+      # rename columns to with shuffled_ prefix for code clarity
+      dplyr::rename(
+        shuffled_or_bulk = or_bulk, 
+        shuffled_or_pseudobulk = or_pseudobulk
+      ) 
+  }) |> purrr::list_rbind(names_to = "replicate")
+```
+
+
+Below, we visualize the shuffled odds ratio distributions, with a red line representing the odds ratio calculated from the real data.
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_bulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_bulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the bulk direction", 
+    x = "Shuffled bulk-direction odds ratios"
+  )
+```
+
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_pseudobulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_pseudobulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the pseudobulk direction", 
+    x = "Shuffled pseudobulk-direction odds ratios"
+  )
+```
+
+
+Next, we can calculate P-values for each gene set.
+We'll also multiply each final P-value by 2 since we are performing 2 tests per cell type.
+
+```{r}
+permutation_pvalues <- geneset_names |>
+  purrr::set_names(geneset_names) |>
+  purrr::map(
+    \(geneset_name) {
+      
+    # This will have one row for the given geneset
+    geneset_baseline_df <- baseline_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name)
+    
+    # Define these variables for code clarity
+    baseline_bulk_or <- geneset_baseline_df$or_bulk
+    baseline_pseudobulk_or <- geneset_baseline_df$or_pseudobulk
+      
+    shuffled_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name) |>
+      # Calculate P-value as fraction of shuffled ORs >= baseline OR
+      dplyr::summarize(
+        bulk_pvalue = sum(shuffled_or_bulk >= baseline_bulk_or) / params$reps, 
+        pseudobulk_pvalue = sum(shuffled_or_pseudobulk >= baseline_pseudobulk_or) / params$reps
```

Comment:
Oops, meant to say this looks correct as well – as or more extreme than the actual OR

Reply:
Somewhere in here you may also want to set a minimum pvalue of 1/reps (or 1/(reps+1), depending how pedantic you want to be). Otherwise FDR won't properly reflect corrections for "most extreme" values.

Reply:
@jashapiro I'm not familiar with this actually? I've done comparatively less fdr correction in my day vs bonferroni or such... 
Here's what I interpret this to mean:

```
      dplyr::mutate(
        q_value = p.adjust(p_value_uncorrected, method = "BH"), 
        q_value = dplyr::case_when(
          q_value > 1 ~ 1, 
          #### this line:
          q_value < 1/(params$reps+1) ~ 1/(params$reps+1),
          .default = q_value
     ))
```
Where `params$reps` is 1000 by default. Is this the idea?

Reply:
Good point/catch, @jashapiro. I thought about that and then promptly forgot. The point is to set the minimum p-value as Josh describes _before you correct_, @sjspielman

Reply:
Aha, so something like this (pseudocode) where `threshold = 1(params$reps)+1)`?

```
mutate(p = ifelse(p < threshold, threshold, p), 
       q = p.adjust(p, method = "BH") # p.adjust already upper bounds at 1
)
```

Reply:
That looks right to me.

Reply:
> Aha, so something like this (pseudocode) where `threshold = 1/(params$reps+1)`?
> 
> ```
> mutate(p = ifelse(p < threshold, threshold, p), 
>        q = p.adjust(p, method = "BH") # p.adjust already upper bounds at 1
> )
> ```

Yes, but I would use `max(p, threshold)`

Reply:
A bit of further explanation: it isn't about the FDR per se as much as just calculating p values from permutations. Since the real p value is never 0, you can set the minimum as if the _next_ permutation would have passed whatever threshold. To be more exact/conservative, you could apply this always, to set all p values as  (n_extreme + 1)/(n_perm + 1).